### ISSUE TITLE
Utility for Safely Inferring Strong Types

### ISSUE THREAD

**Issue by sno2** [microsoft/TypeScript#48232]
# Suggestion

When writing conditional types, having to do extra asserts on inferred types is just as common as another blade of grass on the lawn.  However, this problem does not have to last forever.  My proposal is to add a `TypedInfer` utility type to allow for strongly asserting that a type meets your given constraints:

```ts
type TypedInfer<T, K extends T> = K;
```

You would use it like this:

```ts
type TakeString<T extends string> = T;
type Foo = "Hello" extends TypedInfer<string, infer $Value> ? TakeString<$Value>
  : never;

type TakeStrings<First extends string, Rest extends string[]> = [First, Rest];
type Foo2 = ["Hello", "Awesome", "World"] extends
  [TypedInfer<string, infer $Value>, ...TypedInfer<string[], infer $Rest>]
  ? TakeStrings<$Value, $Rest>
  : never;
```

Instead of having to write this:

```ts
type TakeString<T extends string> = T;
type Foo = "Hello" extends infer $Value
  ? TakeString<$Value extends string ? $Value : never>
  : never;

type TakeStrings<First extends string, Rest extends string[]> = [First, Rest];
type Foo2 = ["Hello", "Awesome", "World"] extends [infer $First, ...infer $Rest]
  ? TakeStrings<
    $First extends string ? $First : never,
    $Rest extends string[] ? $Rest : never
  >
  : never;
```

Also note that it is extremely hard to use `TypedInfer` incorrectly because it won't even compile if you mismatch the `infer` part and the pattern:

```ts
type f = "hello" extends TypedInfer<infer $Value, string> ? true : false;
```

```
error: TS2344 [ERROR]: Type 'string' does not satisfy the constraint '$Value'.
  '$Value' could be instantiated with an arbitrary type which could be unrelated to 'string'.
type f = "hello" extends TypedInfer<infer $Value, string> ? true : false;
```

## 🔍 Search Terms

strong infer utility, infer as type, assert inferred type's type

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## 📃 Motivating Example

Currently, when you `infer` something, you lose complete context of what it was.  I believe this is great because it forces you to write safe types.  Here is an example of what I mean:

```ts
type Initial<N extends string> =
  //
  N extends `${infer $FirstChar}${string}` ? Capitalize<$FirstChar>
    : never;

type Acronym<Words extends string[], $Acc extends string = ""> =
  //
  Words extends [infer $Word, ...infer $Rest]
    ? Acronym<$Rest, $Acc extends "" ? Initial<$Word> : `${$Acc}${Initial<$Word>}`>
    : $Acc;
```

You expect this code to run, however, because `infer _` loses context of the original types, it does not know that `$Word` can only be `string` and `$Rest` can only be `string[]`.  Therefore, we get the following errors:

```
error: TS2344 [ERROR]: Type '$Rest' does not satisfy the constraint 'string[]'.
  Type 'unknown[]' is not assignable to type 'string[]'.
    Type 'unknown' is not assignable to type 'string'.
    ? Acronym<$Rest, $Acc extends "" ? $Word : `${$Acc}${Initial<$Word>}`>
              ~~~~~

TS2344 [ERROR]: Type '$Word' does not satisfy the constraint 'string'.
    ? Acronym<$Rest, $Acc extends "" ? $Word : `${$Acc}${Initial<$Word>}`>
                                                                 ~~~~~
```

Now, to fix this code, we can add checks to the inferred values:

```ts
type Initial<N extends string> =
  //
  N extends `${infer $FirstChar}${string}` ? Capitalize<$FirstChar>
    : never;

type Acronym<Words extends string[], $Acc extends string = ""> =
  //
  Words extends [`${infer $Word}`, ...infer $Rest] ? Acronym<
    $Rest extends string[] ? $Rest : never,
    $Acc extends "" ? Initial<$Word> : `${$Acc}${Initial<$Word>}`
  >
    : $Acc;
```

Our code is now compiling.  Also, the reason I did not add another check for `$Word` is because you can type the inferred patterns of `string`s by wrapping it in a template string.  However, this trick does not work for other types besides `string`.

Anyways, adding these extra checks makes our code obtuse and seem buggy with all that `never`.  It also seems like the conditional clause may have false-positive passes and that's why we're adding the conditional checks to assert the types.  To alleviate these issues, the `TypedInfer` type would allow us to assert the types of the items _while_ inferring them without adding any `never` checks.

```ts
type TypedInfer<T, K extends T> = K;

type Initial<N extends string> =
  //
  N extends `${infer $FirstChar}${string}` ? Capitalize<$FirstChar>
    : never;

type Acronym<Words extends string[], $Acc extends string = ""> =
  //
  Words extends
    [TypedInfer<string, infer $Word>, ...TypedInfer<string[], infer $Rest>]
    ? Acronym<
      $Rest,
      $Acc extends "" ? Initial<$Word> : `${$Acc}${Initial<$Word>}`
    >
    : $Acc;
```

## 💻 Use Cases

I want to use this in my code because I am working on a full type arithmetic library using `string`s for values with n-sized `string` algorithms and using this `TypedInfer` utility has greatly decreased my need to assert the types of various things.  I have written a lot of complex types and the number of times I have had these extra conditional checks to assert the type of an inferred type is way too large.

**Comment by sno2** [comment]
There could also be lints added for using `TypedInfer` incorrectly like the following:

```ts
type f = "hello" extends TypedInfer<infer $Value, string> ? true : false;
```

**Comment by MartinJohns** [comment]
If I understand correctly, your issue would be solved by #48112 as well.

**Comment by sno2** [comment]
Superseded by #48112
